File: src/main/java/net/yakclient/mixin/internal/loader/ClassTarget.java
package net.yakclient.mixin.internal.loader;

import net.yakclient.mixin.internal.bytecode.ByteCodeUtils;

import java.util.Objects;

public class ClassTarget extends PackageTarget {
    private final String cls;

    public ClassTarget(String[] path, String cls) {
        super(path);
        this.cls = cls;
    }

    public static ClassTarget create(String path) {
        try {
            if (ByteCodeUtils.classExists(path)) {
                final String[] fromPath = PackageTarget.fromPath(path);

                final int i = fromPath.length - 1;
                final String[] split = new String[i];

                System.arraycopy(fromPath, 0, split, 0, i);

                return new ClassTarget(split, fromPath[i]);
            } else throw new ClassNotFoundException("Failed to find specified class");
        } catch (ClassNotFoundException e) {
            throw new IllegalArgumentException("Invalid class path");
        }
    }

    public static ClassTarget create(Class<?> path) {
        final String[] fromPath = PackageTarget.fromPath(path.getName());

        final int i = fromPath.length - 1;
        final String[] split = new String[i];

        System.arraycopy(fromPath, 0, split, 0, i);

        return new ClassTarget(split, fromPath[i]);
    }

    public PackageTarget toPackage() {
        return new PackageTarget(this.path);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        ClassTarget that = (ClassTarget) o;
        return Objects.equals(cls, that.cls);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), cls);
    }

    @Override
    public String toString() {
        return super.toString() + "." + this.cls;
    }
}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/loader/ClassTarget.kt
package net.yakclient.mixin.base.internal.loader

import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils.classExists
import java.util.*

class ClassTarget(path: List<String>, private val cls: String) : PackageTarget(path) {
    fun toPackage(): PackageTarget {
        return PackageTarget(path)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        if (!super.equals(other)) return false
        val that = other as ClassTarget
        return cls == that.cls
    }

    override fun hashCode(): Int {
        return Objects.hash(super.hashCode(), cls)
    }

    override fun toString(): String {
        return super.toString() + "." + cls
    }

    companion object {
        @JvmStatic
       fun of(path: String): ClassTarget {
            return try {
                if (classExists(path)) {
                    val fromPath: List<String> = fromPath(path)
                    val i = fromPath.size - 1
                    val split = fromPath.subList(0, i);
//                    val split = arrayOfNulls<String>(i)
//                    System.arraycopy(fromPath, 0, split, 0, i)
                    ClassTarget(split, fromPath[i])
                } else throw ClassNotFoundException("Failed to find specified class")
            } catch (e: ClassNotFoundException) {
                throw IllegalArgumentException("Invalid class path")
            }
        }
        @JvmStatic

        fun of(path: Class<*>): ClassTarget {
            val fromPath: List<String> = fromPath(path.name)
            val i = fromPath.size - 1
            val split = fromPath.subList(0, i);
            return ClassTarget(split, fromPath[i])
        }
    }
}